### DISCUSSION TITLE
How do you split fonts into multiple files with `unicode-range`?

### DISCUSSION THREAD

**Discussion by vnphanquang** [vuejs/vitepress#1648] [General]
I was searching the web for ways to optimize fonts and thought hmm, let's see how the vue/vite eco does it (always a good inspiration pool). I [see here](https://github.com/vuejs/vitepress/tree/main/src/client/theme-default/fonts) fonts are split into subsets with different [unicode ranges](https://github.com/vuejs/vitepress/blob/main/src/client/theme-default/styles/fonts.css). This is very good imo for most of my practical use cases.

Can anyone share the strategy for how to do this? Is there a dedicated tool? Any caveats?

Thank you!

**Comment by brc-dd** [comment]
I'd used https://github.com/fonttools/fonttools for that. This command to be specific: https://fonttools.readthedocs.io/en/latest/subset/index.html

Refer https://stackoverflow.com/a/49755188/11613622, you can also refer the other answer (it's an npm package instead of a python one, haven't tried it though).

Not sure about the caveats though, but IMO should always save the bandwidth unless one is using all sorts of unicode ranges in their app. Might not work (properly) in legacy browsers: https://caniuse.com/font-unicode-range

**Reply by vnphanquang** [reply]
Thank you @brc-dd for the quick and detailed response. Very helpful indeed.

**Reply by vnphanquang** [reply]
@brc-dd I was browsing the vue docs site today (on a linux machine) and found out texts was actually rendered in `Roboto` and not `Inter`, even though `Inter` was loaded. Vite, Vitepress, Vitest docs are actually okay, however.

I didn't see anything particularly off though, so not sure why, just thought you should know.

**Reply by brc-dd** [reply]
Not sure, looks like an issue with https://github.com/vuejs/theme. Some parts are being rendered in Inter and some are in system fonts.

![image](https://user-images.githubusercontent.com/40380293/205487633-dc132674-0eb6-4c6f-a7fe-095b156445f1.png)

**Reply by vnphanquang** [reply]
Yeah, I have no clue. Not a huge impact though.

**Reply by brc-dd** [reply]
Seems like their font file has no glyph for 400 weight. For rest of the weights it's being applied properly.

**Reply by vnphanquang** [reply]
Ah, makes sense. That explains why [inter-latin](https://vuejs.org/assets/inter-latin.7b37fe23.woff2) fetched from vuejs.org is just 36kb, way smaller than what I expect. Thanks for investigating @brc-dd

**Reply by brc-dd** [reply]
I'm not exactly sure if 400 is actually missing as it is a variable font (it's working even for 399 and 401), something is messed up with font-face declarations probably.

**Reply by vnphanquang** [reply]
I tried replacing the new latin subset in vuejs.org (in my localhost) with one extracted using with the `woff2` web variable downloaded from https://rsms.me/inter, which worked fine. So yeah maybe something went wrong during the `pyftsubset` operation.

**Reply by brc-dd** [reply]
The fonts present on vuejs/theme repo weren't generated using pyftsubset. They were probably downloaded from Google Fonts itself. If I remove all font-faces there except the first one and remove the unicode-range, even then everything works fine (with those same font files).
